contracts/helper/RebalancingSetCTokenBidder.sol

```diff
@@ -0,0 +1,357 @@
+            ERC20Wrapper.approve(
+                cTokenAddress,
+                address(_transferProxy),
+                CommonMath.maxUInt256()
+            );
+        }
+    }
+
+    /* ============ External Functions ============ */
+
+    /**
+     * Bid on rebalancing a given quantity of sets held by a rebalancing token wrapping or unwrapping
+     * a target cToken involved. The tokens are returned to the user.
+     *
+     * @param  _rebalancingSetToken    Instance of the rebalancing token being bid on
+     * @param  _quantity               Number of currentSets to rebalance
+     * @param  _allowPartialFill       Set to true if want to partially fill bid when quantity
+                                       is greater than currentRemainingSets
+     */
+
+    function bidAndWithdraw(
+        IRebalancingSetToken _rebalancingSetToken,
+        uint256 _quantity,
+        bool _allowPartialFill
+    )
+        external
+        nonReentrant
+    {
+        // Get token addresses
+        address[] memory combinedTokenArray = _rebalancingSetToken.getCombinedTokenArray();
+
+        // Get inflow and outflow arrays for the given bid quantity
+        uint256[] memory inflowArray;
+        uint256[] memory outflowArray;
+        (
+            inflowArray,
+            outflowArray
+        ) = _rebalancingSetToken.getBidPrice(_quantity);
+
+        // Ensure allowances and transfer auction tokens or underlying from user
+        depositComponents(
+            combinedTokenArray,
+            inflowArray
+        );
+
+        // Bid in auction
+        rebalanceAuctionModule.bidAndWithdraw(
+            address(_rebalancingSetToken),
+            _quantity,
+            _allowPartialFill
+        );
+
+        // Withdraw auction tokens or underlying to user
+        withdrawComponentsToSender(
+            combinedTokenArray
+        );
+
+        // Log bid placed with Eth event
+        emit BidPlacedCToken(
+            address(_rebalancingSetToken),
+            msg.sender
+        );
+    }
+
+    /*
+     * Get token inflows and outflows and combined token array denominated in underlying required
+     * for bid for a given rebalancing Set token.
+     *
+     * @param _quantity               The amount of currentSet to be rebalanced
+     * @param _rebalancingSetToken    The rebalancing Set Token instance
+     * @return inflowArray            Array of amount of tokens inserted into system in bid
+     * @return outflowArray           Array of amount of tokens returned from system in bid
+     * @return combinedTokenArray     Array of token addresses
+     */
+    function getBidPriceAndAddressArray(
+        uint256 _quantity,
+        IRebalancingSetToken _rebalancingSetToken
+    )
+        public
+        view
+        returns (uint256[] memory, uint256[] memory, address[] memory)
+    {
+        // Get token addresses
+        address[] memory combinedTokenArray = _rebalancingSetToken.getCombinedTokenArray();
```

Comment:
To reduce duplication in this method and the above, you can create an internal function called `getTokenFlows` that returns the addresses + inflow./outflow